### ISSUE TITLE
`ActionController::Parameters#require` signature needs adjustment

### ISSUE THREAD

**Issue by andyw8** [Shopify/rbi-central#231]
See https://github.com/Shopify/rbi-central/pull/224#pullrequestreview-1917467194

**Comment by amomchilov** [comment]
@aprescott Do you know what the full possible set of returnable types is?

It should be `T.any(String, Integer, T::Array[T.any(String, Integer))` at a minimum, but do you know of any other possible cases?

Whatever that type is, that's what should be used as the return value of not just `#require`, but also `#[]`, which still [returns `T.untyped`](https://github.com/bdewater/rbi-central/blob/c8b5f1a749af856dc3d79fdaeb7b4d74b13f687c/rbi/annotations/actionpack.rbi#L111-L112)

> although interestingly it doesn't always fire runtime errors, as in cases like a `T::Array[X]` vs. `T::Array[Y]`.

That's a performance concession, Sorbet's runtime type-checking doesn't do deep recursive type-checks on Enumerables. The runtime type-checking calls [`#valid?`](https://github.com/sorbet/sorbet/blob/2d91ebab7b642b900e02233bd49f593ad355469d/gems/sorbet-runtime/lib/types/types/typed_array.rb#L20-L23), not [`#recursively_valid?`](https://github.com/sorbet/sorbet/blob/2d91ebab7b642b900e02233bd49f593ad355469d/gems/sorbet-runtime/lib/types/types/typed_enumerable.rb#L37)

**Comment by aprescott** [comment]
> Do you know what the full possible set of returnable types is?

I'm not sure! It would be good to match whatever `Parameters` can possibly support. String or Integer or Array containers of those certainly sounds reasonable. Perhaps Rails source would help confirm. I don't really know much about the rbi-central project's approach here, but perhaps `T.anything` is an option if the full set of output types isn't clear? The true correct typing would of course beat `T.anything`.

> Sorbet's runtime type-checking doesn't do deep recursive type-checks on Enumerables

Found the docs explaining this right after my comment, realized my confusion!

**Comment by amomchilov** [comment]
> String or Integer or Array containers of those certainly sounds reasonable.

Agreed, let's start there, and adjust it if people find that other values were possible.

@aprescott Would you be interested in opening your first RBI Central PR to contribute this Sig? :)

> I don't really know much about the rbi-central project's approach here, but perhaps `T.anything` is an option if the full set of output types isn't clear?

I like using `T.anything` in my own code, but it's not a good fit for RBI central, because:

1. It gets in peoples' way. Personally I find that to be a good thing, to make me think about the types and act accordingly. But others see it as an intrusive hassle. Ruby developers are already used to dealing with unknown types, anyway.
2. It's kicking the can down the road. RBI Central is supposed to do the homework to provide the types, and `T.anything` largely indicates "🤷 we don't even know, either"

**Comment by duelinmarkers** [comment]
Since anything can be written into params, `T.anything` and `T.untyped` seem like the only correct candidates.

```ruby
>> p = ActionController::Parameters.new
=> #<ActionController::Parameters {} permitted: false>
>> p[:foo] = Struct.new(:a, :b).new("Aye", "Bee")
=> #<struct  a="Aye", b="Bee">
>> p["foo"]
=> #<struct  a="Aye", b="Bee">
>> p.require("foo")
=> #<struct  a="Aye", b="Bee">
```

[This note on `T.anything` and RBIs](https://sorbet.org/docs/anything#tanything-and-rbis) suggests to me that `T.untyped` would be best here, but I don't have a lot of context on the design of annotations.
